ludwig/models/llm.py

```diff
@@ -66,6 +91,34 @@ def __init__(
 
         clear_data_cache()
 
+    def to_device(self, device):
+        device = torch.device(device)
+        if device == self.curr_device:
+            logger.warning(f"LLM already on device '{device}'. Skipping device placement step.")
+            return self
+        else:
+            logger.info(f"Moving LLM from '{self.curr_device}' to '{device}'.")
+
+        model_kwargs = {}
+        if device == torch.device("cuda"):
+            num_gpus = torch.cuda.device_count()
+            model_kwargs.update(
+                dict(
+                    low_cpu_mem_usage=True,
+                    torch_dtype=torch.float16,
+                    device_map="auto",
+                    max_memory={i: "13GiB" for i in range(num_gpus)},
+                )
+            )
+
+        # we save and reload the weights to ensure that they can be sharded across the GPUs using `from_pretrained`
+        with tempfile.TemporaryDirectory() as tmpdir:
+            self.model.save_pretrained(tmpdir)
+            self.model = AutoModelForCausalLM.from_pretrained(tmpdir, **model_kwargs)
```

Comment:
A couple of questions/thoughts here:

1. We should only do this if `torch.device` is CUDA, so I'd nest it in the if condition since we can also have device as `mps` in which case I don't think we need to reload like this? In the case of MPS, we might just be able to `self.model.to(device)`. 
2. Is it necessary to save the pretrained model and then load it back in? If I understand the motivation for saving the pretrained model to disk, it's to prevent downloading the files from hf hub again. If that's the case, then we can actually get away without doing this because huggingface maintains its own cache where the downloaded files are saved and it defaults to checking the cache first before downloading from hf hub, so it should be roughly the same performance wise to just do `AutoModelForCausalLM.from_pretrained(self.model_name, **kwargs).

Reply:
Ah, so I had left a comment that may have not been pushed. The reason we save it and load it back in is so that we can  place it onto the new device, since HF's `from_pretrained` implements device placement in a simple way that pytorch does not (see torch docs for their very low-level approach: https://pytorch.org/tutorials/intermediate/model_parallel_tutorial.html)

Reply:
In any case, I've added the `else` code path to ensure we call `self.model.to(device)` for non-CUDA device placements.